packages/substrate-codegen/src/view-builder/view-builder.ts

```diff
@@ -0,0 +1,254 @@
+    _accountId: WithShapeWithoutMeta<AccountIdDecoded>,
+  ): ShapedDecoder => {
+    if (input.type === "primitive") return primitives[input.value]
+    if (input.type === "compact")
+      return input.isBig ? primitives.compactBn : primitives.compactNumber
+    if (input.type === "bitSequence") return primitives.bitSequence
+
+    if (
+      input.type === "sequence" &&
+      input.value.type === "primitive" &&
+      input.value.value === "u8"
+    ) {
+      return primitives.Bytes
+    }
+
+    const buildNext = (nextInput: LookupEntry): ShapedDecoder => {
+      if (!stack.has(nextInput)) {
+        const nextStack = new Set(stack)
+        nextStack.add(input)
+        const result = _buildShapedDecoder(
+          nextInput,
+          lookupData,
+          nextStack,
+          circularOnes,
+          _accountId,
+        )
+        if (circularOnes.has(input)) {
+          circularOnes.get(input)!.shape = result.shape
+          circularOnes.set(input, result)
+        }
+        return result
+      }
+
+      const result = selfDecoder(() => circularOnes.get(input)!)
+      circularOnes.set(input, result)
+
+      return result
+    }
+
+    const buildVector = (inner: LookupEntry, len?: number) => {
+      const _inner = buildNext(inner)
+      return len ? complex.Array(_inner, len) : complex.Sequence(_inner)
+    }
+
+    const buildTuple = (value: LookupEntry[]) =>
+      complex.Tuple(...value.map(buildNext))
+
+    const buildStruct = (value: StringRecord<LookupEntry>) =>
+      complex.Struct(mapStringRecord(value, buildNext))
+
+    if (input.type === "array") {
+      // Bytes case
+      if (input.value.type === "primitive" && input.value.value === "u8") {
+        return input.len === 32 && (input.id === 0 || input.id === 1)
+          ? _accountId
+          : primitives.BytesArray(input.len)
+      }
+
+      return buildVector(input.value, input.len)
+    }
+
+    if (input.type === "sequence") return buildVector(input.value)
+    if (input.type === "tuple") return buildTuple(input.value)
+    if (input.type === "struct") return buildStruct(input.value)
+
+    // it has to be an enum by now
+    const dependencies = Object.values(input.value).map((v) => {
+      if (v.type === "primitive") return primitives._void
+      if (v.type === "tuple" && v.value.length === 1) {
+        return buildNext(v.value[0])
+      }
+      return v.type === "tuple" ? buildTuple(v.value) : buildStruct(v.value)
+    })
+
+    const inner = Object.fromEntries(
+      Object.keys(input.value).map((key, idx) => [key, dependencies[idx]]),
+    ) as StringRecord<ShapedDecoder>
+
+    const indexes = Object.values(input.value).map((x) => x.idx)
+    const areIndexesSorted = indexes.every((idx, i) => idx === i)
+
+    return areIndexesSorted
+      ? complex.Enum(inner)
+      : complex.Enum(inner, indexes as any)
+  },
+)
+
+export const getViewBuilder: GetViewBuilder = (metadata: V14) => {
```

Comment:
In addition to the experiment, I would add unit tests.
We could probably use [snapshots](https://vitest.dev/guide/snapshot.html).

It would be interesting to test `utility.batch` or `utility.batchAll` calls

Reply:
I added a couple of snapshot tests: one of them is for a `batchCall`, we should add more tests later on. For sure!